Fabric.Identity.API/scripts/Install-Identity-Utilities.psm1

```diff
@@ -0,0 +1,642 @@
+        $keyPath = [io.path]::combine($keyFolder, $keyname)
+
+        if ([io.file]::exists($keyPath))
+        {        
+            $acl = Get-Acl $keyPath
+            $acl.AddAccessRule($allowRule)
+            Set-Acl $keyPath $acl -ErrorAction Stop
+            Write-DosMessage -Level "Information" -Message "The permission '$($permission)' was successfully added to the private key for user '$($iisUser)'"
+        }else{
+            Write-DosMessage -Level "Error" -Message "No key file was found at '$($keyPath)' for '$($signingCert)'. Ensure a valid signing certificate was provided"
+            throw
+        }
+    }catch{
+        Write-DosMessage -Level "Error" -Message "There was an error adding the '$($permission)' permission for the user '$($iisUser)' to the private key. Ensure you selected a certificate that you have read access on the private key. Error $($_.Exception.Message)."
+        throw
+    }
+}
+
+function Get-AppInsightsKey([string] $appInsightsInstrumentationKey, [bool] $quiet){
+    if(!$quiet){
+        $userEnteredAppInsightsInstrumentationKey = Read-Host  "Enter Application Insights instrumentation key or hit enter to accept the default [$appInsightsInstrumentationKey]"
+
+        if(![string]::IsNullOrEmpty($userEnteredAppInsightsInstrumentationKey)){   
+            $appInsightsInstrumentationKey = $userEnteredAppInsightsInstrumentationKey
+        }
+    }
+    if($appInsightsInstrumentationKey){ Add-InstallationSetting "identity" "appInsightsInstrumentationKey" "$appInsightsInstrumentationKey" | Out-Null }
+    if($appInsightsInstrumentationKey){ Add-InstallationSetting "common" "appInsightsInstrumentationKey" "$appInsightsInstrumentationKey" | Out-Null }
+    return $appInsightsInstrumentationKey
+}
+
+function Get-SqlServerAddress([string] $sqlServerAddress, [bool] $quiet){
+    if(!$quiet){
+        $userEnteredSqlServerAddress = Read-Host "Press Enter to accept the default Sql Server address '$($sqlServerAddress)' or enter a new Sql Server address" 
+
+        if(![string]::IsNullOrEmpty($userEnteredSqlServerAddress)){
+            $sqlServerAddress = $userEnteredSqlServerAddress
+        }
+    }
+    if($sqlServerAddress){ Add-InstallationSetting "common" "sqlServerAddress" "$sqlServerAddress" | Out-Null }
+    return $sqlServerAddress
+}
+
+function Get-IdentityDatabaseConnectionString([string] $identityDbName, [string] $sqlServerAddress, [bool] $quiet){
+    if(!$quiet){
+        $userEnteredIdentityDbName = Read-Host "Press Enter to accept the default Identity DB Name '$($identityDbName)' or enter a new Identity DB Name"
+        if(![string]::IsNullOrEmpty($userEnteredIdentityDbName)){
+            $identityDbName = $userEnteredIdentityDbName
+        }
+    }
+    $identityDbConnStr = "Server=$($sqlServerAddress);Database=$($identityDbName);Trusted_Connection=True;MultipleActiveResultSets=True;"
+
+    Invoke-Sql $identityDbConnStr "SELECT TOP 1 ClientId FROM Clients" | Out-Null
```

Comment:
If the DbName is changed and they don't accept the default. If they type it in wrong Database the doesn't exist, will there be a failure on the Invoke-Sql that is not caught?

Reply:
Invoke-Sql internally has a try catch that will catch the error, write out the error, and then forcibly stop the script using the `-ErrorAction Stop` parameter. The Out-Null is simply used here to suppress extraneous output from the successful execution of the sql statement.